### PR TITLE
A 791/replay subject fix with connect

### PR DESCRIPTION
**Pull request by mineme0110**

### CODE REVIEW

File: scalanet/src/io/iohk/scalanet/peergroup/StaticAddressMappedPeerGroup.scala [hunk 1]
```diff
@@ -21,23 +23,27 @@ class StaticAddressMappedPeerGroup[A, AA, M](
 ) extends PeerGroup[A, M] {
 
   private val reverseLookup = config.knownPeers.map(_.swap)
-
+  implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/StaticAddressMappedPeerGroup.scala]: would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/SimplePeerGroup.scala [hunk 2]
```diff
@@ -59,51 +61,64 @@ class SimplePeerGroup[A, AA, M](
...
-    underLyingPeerGroup
-      .server()
-      .collect(ChannelCreated.collector)
-      .mergeMap(channel => channel.in)
-      .collect {
-        case Left(e: EnrolMe[A, AA]) => e
-      }
-      .foreach(handleEnrollment)
+    try {
+      implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/SimplePeerGroup.scala]: would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala [hunk 3]
```diff
@@ -46,8 +47,11 @@ class TCPPeerGroup[M](val config: Config)(implicit codec: StreamCodec[M], bi: Bu
     extends TerminalPeerGroup[InetMultiAddress, M]() {
 
   private val log = LoggerFactory.getLogger(getClass)
+  implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala]: would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala [hunk 4]
```diff
@@ -48,6 +49,7 @@ class TLSPeerGroup[M](val config: Config)(
 ) extends TerminalPeerGroup[InetMultiAddress, M]() {
 
   private val log = LoggerFactory.getLogger(getClass)
+  implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala]: Same as before, would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala [hunk 5]
```diff
@@ -204,6 +209,7 @@ object TLSPeerGroup {
   ) extends Channel[InetMultiAddress, M] {
 
     private val log = LoggerFactory.getLogger(getClass)
+    implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala]: Same as before, would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala [hunk 6]
```diff
@@ -312,7 +319,10 @@ object TLSPeerGroup {
   ) extends Channel[InetMultiAddress, M] {
 
     private val log = LoggerFactory.getLogger(getClass)
-    private val messageSubject = ReplaySubject[M]()
+    implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/TLSPeerGroup.scala]: Same as before, would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala [hunk 7]
```diff
@@ -37,8 +38,10 @@ class UDPPeerGroup[M](val config: Config)(implicit codec: Codec[M], bufferInstan
     extends TerminalPeerGroup[InetMultiAddress, M]() {
 
   private val log = LoggerFactory.getLogger(getClass)
-
-  private val serverSubject = PublishSubject[ServerEvent[InetMultiAddress, M]]()
+  implicit val scheduler: Scheduler = Scheduler.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala]: Same as before, would it be better to have this as an implicit parameter of the class constructor?

File: scalanet/src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala [hunk 8]
```diff
@@ -111,9 +114,11 @@ object TCPPeerGroup {
       codec: StreamCodec[M],
       bi: BufferInstantiator[ByteBuffer]
   ) extends Channel[InetMultiAddress, M] {
+    import monix.execution.Scheduler.Implicits.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala]: similar as above, implicit that could be taken from parameter

File: scalanet/src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala [hunk 9]
```diff
@@ -66,25 +68,28 @@ class DTLSPeerGroup[M](val config: Config)(
...
+    private val connectableObservable = ConnectableObservable.cacheUntilConnect(channelSubject, PublishSubject[M]())
+
+    override val in: ConnectableObservable[M] = connectableObservable
 
     override def sendMessage(message: M): Task[Unit] = {
       import io.iohk.scalanet.peergroup.BufferConversionOps._
       val buffer = codec.encode(message)
 
       Task
-        .async[Unit] { (_, c) =>
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala]: did the interface change from RC1 to RC3? I could not find how this was working with the pair as function parameter.

File: scalanet/test/src/io/iohk/scalanet/peergroup/InMemoryPeerGroupSpec.scala [hunk 10]
```diff
@@ -61,17 +61,17 @@ class InMemoryPeerGroupSpec extends FlatSpec with BeforeAndAfter {
...
       val aliceClient1 = alice.client(bob.processAddress).evaluated
-      aliceClient1.sendMessage(aliceMessage1).runAsync
+      aliceClient1.sendMessage(aliceMessage1).runToFuture
 
-      val bobServer = bob.server().collectChannelCreated.headL.runAsync
+      val bobServer = bob.server().collectChannelCreated.headL.runToFuture
 
       intercept[Exception] {
         bobServer.futureValue.to shouldBe alice.processAddress
       }
```
Comment (EzequielPostan) [inline scalanet/test/src/io/iohk/scalanet/peergroup/InMemoryPeerGroupSpec.scala]: I believe that this test case can be removed now. It used to test that the message received before subscribing to `server` was supposed to be lost.

File: scalanet/test/src/io/iohk/scalanet/peergroup/StaticAddressMappedPeerGroupSpec.scala [hunk 11]
```diff
@@ -28,17 +27,20 @@ class StaticAddressMappedPeerGroupSpec extends FlatSpec {
       "Bob"
     ) { (alice, bob) =>
       println(s"Alice is ${alice.processAddress}, bob is ${bob.processAddress}")
-      val alicesMessage = Random.alphanumeric.take(1024).mkString
-      val bobsMessage = Random.alphanumeric.take(1024).mkString
+      val alicesMessage = "Hi Bob" //Random.alphanumeric.take(1024).mkString
+      val bobsMessage = "Hi Alice" //Random.alphanumeric.take(1024).mkString
```
Comment (EzequielPostan) [inline scalanet/test/src/io/iohk/scalanet/peergroup/StaticAddressMappedPeerGroupSpec.scala]: minor observation, commented code could be removed

File: scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupSpec.scala [hunk 12]
```diff
@@ -39,11 +39,14 @@ class TCPPeerGroupSpec extends FlatSpec with BeforeAndAfterAll {
   it should "report an error for messaging on a closed channel -- server closes" in
     withTwoRandomTCPPeerGroups[String] { (alice, bob) =>
       val alicesMessage = Random.alphanumeric.take(1024).mkString
-      bob.server().collectChannelCreated.foreachL(channel => channel.close().runAsync).runAsync
+      bob.server().collectChannelCreated.foreach(_.in.connect())
```
Comment (EzequielPostan) [inline scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupSpec.scala]: why is this line needed?
I mean wouldn't you be able to close the channels without this?

**Review by EzequielPostan — CHANGES_REQUESTED** [review]
The code looks good. I added some comments to address. 
Could we change calls to implicit scheduler's instances to implicit class parameters? (See code comments)

As a separate comment (not to address in this PR). It would be nice to have a more idiomatic "connection" code than 
```
bob.server().collectChannelCreated.foreach(_.in.connect())
bob.server().connect()
```

Thank you for the PR @mineme0110, good work!

File: scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala [hunk 13]
```diff
@@ -23,6 +26,12 @@ object InetPeerGroupUtils {
     }
   }
 
+  def toTask[T](f: => Cancelable)(implicit tag: ClassTag[T]): Task[Cancelable] = {
```
Comment (krcz) [inline scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala]: is the `tag` used here?

File: scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupCachetUntilConnectSpec.scala [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import io.iohk.scalanet.peergroup.PeerGroup.ServerEvent._
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.Matchers._
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.concurrent.ScalaFutures._
+import org.scalatest.{BeforeAndAfterAll, FlatSpec}
+
+import scala.concurrent.duration._
+
+class TCPPeerGroupCachetUntilConnectSpec extends FlatSpec with BeforeAndAfterAll {
```
Comment (krcz) [inline scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupCachetUntilConnectSpec.scala]: "Cached" or "Cache"

Comment (mineme0110) [reply scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupCachetUntilConnectSpec.scala]: Cached maybe sounds

File: scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala (hunk 13 above)
Comment (mineme0110) [reply scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala]: It doesn't compile without that since the generic function loses the type, there are 2 toTask function. I will check if any otherway to resolve this

File: scalanet/src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala [hunk 15]
```diff
@@ -156,9 +162,10 @@ object TCPPeerGroup {
     private val activationF = activation.future
     private val deactivation = Promise[Unit]()
     private val deactivationF = deactivation.future
+    import monix.execution.Scheduler.Implicits.global
```
Comment (EzequielPostan) [inline scalanet/src/io/iohk/scalanet/peergroup/TCPPeerGroup.scala]: same as before, why not move this as a class constructor parameter?

File: scalanet/src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala (hunk 9 above)
Comment (mineme0110) [reply scalanet/src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala]: > did the interface change from RC1 to RC3? I could not find how this was working with the pair as a function parameter.

Yes it has changed

**Review by EzequielPostan — APPROVED** [review]

File: scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala (hunk 13 above)
Comment (krcz) [reply scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala]: Adding unused type parameter and implicit argument are not a good way to solve the problem. You should probably rename this function to `cancellableToTask` or `taskFromCancellable`.

Comment (krcz) [reply scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala]: What's more during more careful analysis I found that your implementation is wrong as it never fulfils the task. Let's take a look at example usage from Monix documentation:
```
def delayResult1[A](timespan: FiniteDuration)(thunk: => A): Task[A] =
  Task.create { (scheduler, cb) =>
    val c = scheduler.scheduleOnce(timespan)(cb(Try(thunk)))
    // We can simply return `c`, but doing this for didactic purposes!
    Cancelable(() => c.cancel())
  }
```

You can see that `cb` (callback) function is passed and when the computation is finished it has to be called with the result. Otherwise the Task is never considered finished.

Comment (krcz) [reply scalanet/src/io/iohk/scalanet/peergroup/InetPeerGroupUtils.scala]: Plus third, final point: is that version of function even used anywhere? I have removed it and the project compiles, the tests pass.

**Review by krcz — CHANGES_REQUESTED** [review]

**Comment by zejaco** [comment]
Hey @mineme0110 ! Could you wait with this merge till Monday? I'm finishing my performance tests and I don't want to have scalanet version updated in my tests. Thanks a lot in advance!

**Review by krcz — APPROVED** [review]
Thank you for improving the PR! Great work overall! :)

File: scalanet/src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala [hunk 16]
```diff
@@ -34,6 +34,8 @@ class DTLSPeerGroup[M](val config: Config)(
   private val serverConnector = createServerConnector()
 
   private val channelSubject = PublishSubject[ServerEvent[InetMultiAddress, M]]()
```
Comment (jtownson) [inline scalanet/src/io/iohk/scalanet/peergroup/DTLSPeerGroup.scala]: Can we remove the duplication of `channelSubject` and `connectableObservable` in UDP, DTLS, TCP, TLS. (see for eg https://github.com/input-output-hk/scalanet/blob/A-791/Subjects/scalanet/src/io/iohk/scalanet/monix_subjects/CacheUntilConnectSubject.scala)

File: scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupCachedUntilConnectSpec.scala [hunk 17]
```diff
@@ -0,0 +1,60 @@
+package io.iohk.scalanet.peergroup
```
Comment (jtownson) [inline scalanet/test/src/io/iohk/scalanet/peergroup/TCPPeerGroupCachedUntilConnectSpec.scala:1]: I didn't understand what this file is testing. Is it a feature not covered in the main TCPPeerGroupSpec?

**Review by jtownson — COMMENTED** [review]
Please would you:
-- Address comment about duplication of serverSubject and cacheUntilConnectSubject at all usage points.
-- Merge with develop and update kademlia peer group tests.

**Review by jtownson — APPROVED** [review]
Definitely a step forward.

Note: you might want to merge up the monix 3.0.0 change from develop.
